Add attachments to events

sazanof opened the pull request:
Fix https://github.com/nextcloud/calendar/issues/298

Signed-off-by: Mikhail Sazanov <m@sazanof.ru>

I tried to implement attachment files in events. The parameter is described here https://datatracker.ietf.org/doc/html/rfc5545#section-3.8.1.1

Probably it will be a draft, although in fact everything works.

I decided to do this: there are several ways to add attachments:
1) From the Files application. Then a common link is formed (here I think to make it possible to provide a link not shared, but closed. but so far I have done so).
2) File upload. The file is uploaded to the root directory and a public link is formed to it

Probably here you can use a separate folder "Calendar". And in the future, the place for attachments may grow, but it is more convenient for users to download from a computer, rather than choose from "Files".

I also plan to delete the file itself when deleting the event (if the owner of the event does this action), if it is not needed in the repository.

Well, the loading indication (start - end) it should also be added.

Please tell me: where is the email notification template generated when an event notification arrives (I want to add attachments section)?

I would be grateful if you would take a look and give your comments. Thanks!=)

![image](https://user-images.githubusercontent.com/3595562/171364507-0a5e3a96-644e-4403-9c98-f35cc4a9ccb0.png)
empty

![image](https://user-images.githubusercontent.com/3595562/171364636-7b4abc85-d1f9-44fd-89fe-d1464b4372bc.png)
with files

![image](https://user-images.githubusercontent.com/3595562/171364723-7ec5bdc6-4c68-494a-b58c-866484ae9ba0.png)
file options

ChristophWurst commented:
This PR addresses https://github.com/nextcloud/calendar/issues/298

tcitworld commented:
> Please tell me: where is the email notification template generated when an event notification arrives (I want to add attachments section)?

This is handled by the server: https://github.com/nextcloud/server/blob/master/apps/dav/lib/CalDAV/Schedule/IMipPlugin.php

tcitworld commented on src/services/attachmentService.js:
```diff
@@ -0,0 +1,96 @@
+			}))
+		}
+		else {
+			const url = `/remote.php/dav/files/${dav.userId}/${file.name}`
```
There's the brand new `nextcloud-upload` package from @skjnldsv that might be a better fit: https://github.com/nextcloud/nextcloud-upload.

sazanof replied:
> @sazanof Tests have to be updated for the new `attachments` property.

Tell me, is it enough to add a parameter or do I need to write a full test? If it's a complete test, then I ask for help, since **I've never written tests and don't understand how to write them** (It's a matter of time and effort, I don't argue). =( 
I tried to write my own block for checking (by analogy with other tests) `it()`
This error occurs if I add my parameters to 
```
expect(getDefaultEvent Object()).to Equal({
....
attachments: [
	'ATTACHMENT1',
	'ATTACHMENT2',
	'ATTACHMENT3',
	'ATTACHMENT4',
	]
})
```
https://gist.github.com/sazanof/ecb2aaa8c760f93e74d2952294ba09ba

When checking, I get an error. I can't understand why it occurs...
```
 ● Test suite failed to run

    TypeError: Converting circular structure to JSON
        --> --> --> --> --> --> --> --> starting at object with constructor 'CalendarComponent'
        --- property '_root' closes the circle
        at stringify (<anonymous>)

```

skjnldsv replied:
Yes, clearly.

sazanof replied:
@skjnldsv **Well, since we are not in a hurry, we can discuss =).** 

I would like to be able to upload to a specific folder + have a callback to provide shared access or something else. Maybe a subscription to some events (loading before, progress, after) ...
By buttons - you need a universal button that can be entered into any standard NC component. Ex.: ListItem,ActionButton, Button and so on. 

It would also be cool to be able to set the template of the uploaded file (before uploading, at the moment and after).

If you make the component universal, then in the future consider using it to upload not only to DAV, but also when writing a letter in the mail, for example...

tcitworld commented on src/store/calendarObjectInstance.js:
```diff
@@ -1350,6 +1350,65 @@ const mutations = {
+	addAttachmentByLocalFile(state, { calendarObjectInstance, file }){
+		
+		const attachment = AttachmentProperty.fromLink(file.url)
+		// hot-fix needed temporary, becase calendar-js has no fileName get-setter
```
Can you open a PR there? https://github.com/nextcloud/calendar-js

tcitworld commented on src/components/Editor/Attachments/AttachmentsList.vue:
```diff
@@ -0,0 +1,222 @@
+					:force-display-actions="true"
+					:title="attachment.fileName">
+					<template #icon>
+						<Paperclip :size="24" />
```
As an improvement, we *should* be able to show the file preview here instead.

sazanof replied:
@tcitworld  That would be really great! However, I don't really know =( yet how to display file previews (it would also be relevant for downloading attachments in the mail application).  

I know that links like " are generated http://nextcloud.dev.loc/index.php/core/preview?fileId=607478&x=250&y=250&forceIcon=0&a=0"  

I read the description here, but there was a question of generating office documents. 

https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/previews_configuration.html 

**I would be grateful for a hint in this direction.**

tcitworld replied:
I did not find a cleaner way than https://github.com/nextcloud/viewer/blob/43e7003b3095bba553d71f59e73aa8ec1dbdff29/src/mixins/PreviewUrl.js#L57-L82

We should have this exacted in a new package.

sazanof replied:
**Thank you very much. I will apply this method.**

That is, do I understand correctly that if I don't have a `collabora + nextcloud office` bundle, then I get previews only for pictures, mp3 and text files? There is also no way to get a preview at the time of download. 

That's all I'm getting at: I probably need to replace the missing images with "something" (file icons, for example).
BTW
There are still such addresses.
`/index.php/apps/theming/img/core/filetypes/x-office-document.svg`
Probably I should also look for how icons are generated according to the mime type.

sazanof replied:
https://github.com/nextcloud/server/blob/fd5f7486da8192b6ebf69ee3e6640cb7662028ca/apps/files/src/views/Sidebar.vue#L310-L337
As I understand it, so far there is no single way (or component) to get a preview or icon

sazanof replied:
In general, I had to use `MANAGED-ID` and `split()` magic =)
Perhaps there are better solutions. But I can only see this one so far.
![image](https://user-images.githubusercontent.com/3595562/172137159-4589d30d-8624-432e-a84b-9eae87d34268.png)

_In general, the topic of the preview is quite ticklish_

tcitworld commented:
> I also plan to delete the file itself when deleting the event (if the owner of the event does this action), if it is not needed in the repository.

Note : this will only affect events deleted through the web UI.

In the future we need to implement CalDAV-Managed Attachments ([RFC 8607](https://datatracker.ietf.org/doc/html/rfc8607)) https://github.com/sabre-io/dav/issues/837 and let the server handle everything instead. We have time though, I'm not aware of any client implementing this (at least not [Evolution](https://gitlab.gnome.org/GNOME/evolution-data-server/-/issues/125), KOrganizer or even Thunderbird - this should interest @devvv4ever btw)

tcitworld commented:
> I decided to do this: there are several ways to add attachments:
>  1. From the Files application. Then a common link is formed (here I think to make it possible to provide a link not shared, but closed. but so far I have done so).
>  2. File upload. The file is uploaded to the root directory and a public link is formed to it
>
> Probably here you can use a separate folder "Calendar". And in the future, the place for attachments may grow, but it is more convenient for users to download from a computer, rather than choose from "Files".

I'm not fond of uploading directly to the user root folder, it might get crowded real quick.

One possibility would be to handle this in the app backend (it would be irrelevant once [RFC 8607](https://datatracker.ietf.org/doc/html/rfc8607) would be implemented in dav server - see above) and put files in user's appdata. However the users wouldn't have access to files apart from accessing them directly from the event.

In the case of Talk the app creates a `Talk` folder in the user root to put attachments. Is that a pattern we should reproduce here with a `Calendar` folder ? Sounds the best to me. Opinions @nextcloud/designers @nickvergessen

Later and with the help of file tracking for RFC 8607 we can improve these files discoverability just like for Talk with a « all files attached to events in this calendar » section.

tcitworld commented:
@sazanof Tests have to be updated for the new `attachments` property.

sazanof commented:
> @sazanof Tests have to be updated for the new `attachments` property.

Tell me, is it enough to add a parameter or do I need to write a full test? If it's a complete test, then I ask for help, since I've never written tests and don't understand how to write them (It's a matter of time and effort, I don't argue). =( 
I tried to write my own block for checking (by analogy with other tests) `it()`
This error occurs if I add my parameters to 
```
expect(getDefaultEvent Object()).to Equal({
....
attachments: [
	'ATTACHMENT1',
	'ATTACHMENT2',
	'ATTACHMENT3',
	'ATTACHMENT4',
	]
})
```
https://gist.github.com/sazanof/ecb2aaa8c760f93e74d2952294ba09ba

When checking, I get an error. I can't understand why it occurs...
```
 ● Test suite failed to run

    TypeError: Converting circular structure to JSON
        --> --> --> --> --> --> --> --> starting at object with constructor 'CalendarComponent'
        --- property '_root' closes the circle
        at stringify (<anonymous>)
```

devvv4ever commented:
@tcitworld Thank you for thinking on us here too :) We would definitely love to implement attachment sync in DAVx⁵! Managed Attachments would be a really really cool feature also on Android. The problem here is that the Android Calendar storage does not offer attachments storage capabilities. So we'd need to find other apps that can provide an interface for it. We need a collaboration with others then, too (etar, acalendar+, jtx board, etc) because DAVx⁵ is not a calendar app itself so that people can also access their attachments then on Android.

nickvergessen commented:
I think it shouldn't be in the root anymore. I think Deck and etc. use an Attachment/ root folder

nimishavijay commented:
Dunno if the frontend is ready for reviewing but I have some design feedback that could be considered, feel free to incorporate it :) The general flow looks really great! The only feedback is regarding wording and icons.

In the "empty" screenshot:
- The empty content icon could be the [`create-new-folder` MDI icon](https://fonts.google.com/icons?selected=Material%20Icons%3Acreate_new_folder%3A) or just the [`folder` MDI icon](https://fonts.google.com/icons?selected=Material%20Icons%3Afolder%3A)
- Color of the empty content icon should be `var(color-text-maxcontrast)`
- "Choose" --> "Add from Files"
- Upload icon --> [`file-upload` MDI icon](https://fonts.google.com/icons?selected=Material%20Icons%3Afile_upload%3A)

In the "with files" screenshot:
- Instead of a 3dot menu icon that opens a menu with "Choose" and "Upload" we could make it just another list item with an [`add` icon](https://fonts.google.com/icons?selected=Material%20Icons%3Aadd%3A) and text "New attachment"
- If it is possible to show a preview thumbnail we could do that
- If it is possible to display a filetype icon we could do that 
- Currently the subline shows the type of file it is, but that is already understood from the extension in the name of the file. Instead we could show the size of the file (for eg. 300 KB)
- "Choose" --> "Add from Files"
- Upload icon --> [`file-upload` MDI icon](https://fonts.google.com/icons?selected=Material%20Icons%3Afile_upload%3A)

In the "file options" screenshot:
- Not sure if this is already the case but clicking on the file should open it in the viewer if possible. If not it could be downloaded.
- Hence we would not need the "View file" option in the action menu. It could be changed to "Download". Icon stays the same
- "Delete file" --> "Delete"
- Delete icon --> [`delete` MDI icon](https://fonts.google.com/icons?selected=Material%20Icons%3Adelete%3A)

Super nice work! cc @jancborchardt

sazanof commented:
@nimishavijay Hello!

Thank you for the detailed answer and your wishes. I will definitely take them into account when updating the design. 

Regarding the preview for attachments: I have already implemented this, I just closed the discussion, and this branch is no longer shown. My joint.=) Here is a link to this comment 

https://github.com/nextcloud/calendar/pull/4251#discussion_r889991161    

file options actions, in fact, will not be needed if you open a file by clicking on its name and add only one delete button to the right of it. 

But what if we need to introduce a document preview function? Preview in the modal window, download and delete.  Of course it won't be soon, and then you can fix the design, but still. 

Thanks again for your comments!

jancborchardt commented:
Good stuff! Agree with @nimishavijay's feedback, just 2 points/questions:
- To keep keeping it simple, we can put the attachment section on the bottom of "Details" instead of adding a separate "Attachments" section. Details is visible by default.
- As to where the Files are stored: Regardless of how we do it, it would mean that only 1 copy exists and collaboration is possible, right? Or will files uploaded from the cloud be duplicated?

nimishavijay commented:
Just checked Google on how attachments are handled and they are uploaded into the root directory of your Google Drive, and the attachment basically links to that file. If you email another person the the event, they are able to see the attachment and if they click on it, it opens the file in Drive. 

cc @jancborchardt

sazanof commented:
@nimishavijay, @jancborchardt 
I tried to take into account all your comments and added a little of my own. What came out of it - you can see in the screenshots and https://github.com/nextcloud/calendar/pull/4251/commits/a4e764c82cb4774c9ffb4a2f2b742bccd7a53e44.

![image](https://user-images.githubusercontent.com/3595562/172660811-3ffd2a1a-5be9-45fb-a12c-2c280e601ff4.png)
_no attachments_

![image](https://user-images.githubusercontent.com/3595562/172660747-f2fbb1ab-a5c9-4ab6-bee7-6a9cb00d8816.png)
_with attachments_

What do you say?

jancborchardt commented:
@sazanof really cool! :) Some additional feedback:
- [ ] "Upload from PC" → better "Upload from device" since people might be on tablet or mobile
- [ ] The preview needs some border radius (like in Files), `var(--border-radius)`
- [ ] The subline is too technical and too much information, it can just be cut

jancborchardt reviewed (changes requested):
@sazanof really cool! :) Some additional feedback:
- [x] "Upload from PC" → better "Upload from device" since people might be on tablet or mobile
- [x] The preview needs some border radius (like in Files), `var(--border-radius)`
- [ ] The subline is too technical and too much information, it can just be cut

jancborchardt commented:
@sazanof could you add a screenshot for easier design review? Thank you! :)

sazanof commented:
> @sazanof could you add a screenshot for easier design review? Thank you! :)

No problem =)
![image](https://user-images.githubusercontent.com/3595562/174161233-a302f38c-6d27-463d-920d-d010f12aae9f.png)

jancborchardt commented:
@sazanof really nice! Only 2 things actually:
- Better than the trash can icon would be the simple remove "x", cause otherwise people who added files from the cloud could misunderstand it as deletion from there. But it's just a removal. :)
- The indentation of the text is a bit further in than the text below, of xourse cause of the nice thumbnails. Not sure if anything needs to be done about it actually.

What do you think @nimishavijay @marcoambrosini?

IgorA100 commented:
When I create an event from Thunderbird and attach an attachment as a link to a document stored in NC, when I go to the web version of NC, I see the created event, but the link I added is not displayed anywhere.
I consider this a problem that is highly desirable to resolve.
I would be very grateful!

nimishavijay commented:
Looks good once you replace the trash icon :) 🚀

IgorA100 commented:
> PS I really want attachments in the calendar. =)

Super great!
> PSPS @IgorA100 Please send the source of the event

I don't quite understand what I need to do :(

sazanof commented:
> > PS I really want attachments in the calendar. =)
> 
> Super great!
> 
> > PSPS @IgorA100 Please send the source of the event
> 
> I don't quite understand what I need to do :(

an .ics file of an event

sazanof commented:
@jancborchardt @nimishavijay 

https://github.com/nextcloud/calendar/pull/4251/commits/3c62c0ef1ed8ca8dc8fe209d380594f892ac268f

![image](https://user-images.githubusercontent.com/3595562/177539488-eb8ac04e-8b38-44fa-8f31-2c8e92f21d91.png)

What is next? =)

nimishavijay reviewed (approved):
🚀🚀🚀 :)

IgorA100 commented:
> What is next? =)

Show size and change time? :)

sazanof commented:
@IgorA100 
А вы точно используете версию NC с моими изменениями? Потому что если нет, то ваши пропавшие вложения вполне объяснимы =)

Are you sure you're using the NC version with my changes? Because if not, then your missing attachments are quite understandable =)

IgorA100 commented:
> NC с моими изменениями?

Вероятно нет, т.к. я ставил только стандартно через приложения в NC, а там крайняя версия 3.4.1 и в ней вообще нет функции аттача.
Мне нужно собрать из исходников версию "Release 3.4.2-RC1", верно?

sazanof commented:
contact me via telegram @Sazanof   

and these comments can be marked as offtop or deleted , since they are not related to PR

st3iny commented on src/store/calendarObjectInstance.js:
```diff
@@ -1351,6 +1351,61 @@ const mutations = {
+
+		// hot-fix needed temporary, becase calendar-js has no fileName get-setter
+		const parameterFileName = new Parameter('FILENAME', fileName)
+		const managedId = new Parameter('MANAGED-ID', `${sharedData.file_source}||${sharedData.has_preview}`)
```
Custom 
```suggestion
		const managedId = new Parameter('X-MANAGED-ID', `${sharedData.file_source}||${sharedData.has_preview}`)
```

Non-standard parameters should be prefixed with `X-`.

tcitworld replied:
It's https://datatracker.ietf.org/doc/html/rfc8607#section-4.3

st3iny replied:
Oh, right. I'm sorry for my stupidity.

But still, I don't like the custom serialization format. Ideally we should use base64 encoded JSON data and handle the encoding/serialization transparently through calendar-js.

sazanof replied:
@st3iny we can use parameters (X-NC-PREVIEW X-NC-FILE-ID), then this will simplify event processing and their correct selection in the future. now I have found a bug related to the slpit method. I'll redo it.

sazanof replied:
https://github.com/nextcloud/calendar/pull/4251#issuecomment-1207938079

st3iny reviewed (changes requested):
I tested it locally and it works great so far. Huge thanks for implementing this feature!

However, there is a big issue regarding privacy: All attachments are shared publicly by default. Even when I don't share the event with anyone.

For private events, the attachments should not be shared at all (only visible if owner is logged in). When participants are invited we can create a public share + link **but** we need to communicate this to the user.

E.g. Google Calendar explicitly asks for my permission if I invite someone to an event with attachments:

![ksnip_20220712-161609](https://user-images.githubusercontent.com/1479486/178515414-3b2f3ce6-ed77-476e-b9fd-b56d30898d5e.png)

## Translation:
A user requires access to your file

[REDACTED]

**Enable link share**
Everyone with the link will be able to **see** the file

- [ ] Deny access 

[**Cancel**] [**Invite**]

---

If I tick `Deny access` and send the invitation, the attachment link will be present in the ics file but I won't be able to open it in my browser (until I log in as the owner).

st3iny commented on src/components/Editor/Attachments/AttachmentsList.vue:
```diff
@@ -0,0 +1,249 @@
+
+		getPreview(managedId, mime) {
+			if (managedId !== null) {
+				const _mId = managedId.split('||')
+				if ((_mId.length === 2) && (_mId[1] === 'true')) {
+					return generateUrl(`/core/preview?fileId=${_mId[0]}&x=100&y=100&a=0`)
```
See comment from Thomas above.

st3iny commented on src/models/attachment.js:
```diff
@@ -0,0 +1,64 @@
+		formatType: attachmentProperty.formatType,
+		uri: attachmentProperty.uri,
+		value: attachmentProperty.value,
+		managedId: typeof attachmentProperty._parameters.get('MANAGED-ID') !== 'undefined' ? attachmentProperty._parameters.get('MANAGED-ID').value : null,
```
See the comment from Thomas above.

tcitworld reviewed (commented):
I did a small review of the code, still need to test myself.

tcitworld commented:
I'm wondering if https://github.com/nextcloud/server/pull/29886 should be used as an upload folder. :thinking:

szaimen commented:
> I'm wondering if [nextcloud/server#29886](https://github.com/nextcloud/server/pull/29886) could be used as an upload folder. 🤔

I am not sure as well. I mean it would be convenient to be able to work with these files like with any other file too. cc @jancborchardt @nimishavijay

jancborchardt commented:
@szaimen @sazanof @tcitworld yeah, in other apps like Talk, Notes and Collectives we put Files in the filesystem, so probably apt to do it here too.

@miaulalala @ChristophWurst "Calendar" folder, or do you prefer a hidden folder?

miaulalala commented:
> @szaimen @sazanof @tcitworld yeah, in other apps like Talk, Notes and Collectives we put Files in the filesystem, so probably apt to do it here too.
> 
> @miaulalala @ChristophWurst "Calendar" folder, or do you prefer a hidden folder?

I think we should do the same as other apps do.

sazanof commented on src/components/Editor/Attachments/AttachmentsList.vue:
```diff
@@ -0,0 +1,242 @@
+					:key="attachment.path"
+					:force-display-actions="true"
+					:title="getBaseName(attachment.fileName)"
+					@click="openFile(attachment.uri)">
```
PR merged, but nc-vue version didnt' updated in package.json with this change

sazanof commented:
@tcitworld Tomas, I tried to fix everything.

> It's been published

don't understand what did you mean

sazanof commented:
**Oops!! 
I push it again!**

=)

st3iny commented:
@sazanof I rebased, fixed conflicts and am currently testing your PR. I had to force push in order to achieve that.

sazanof commented:
> I rebased, fixed conflicts and am currently testing your PR. I had to force push in order to achieve that. I hope that's not a problem.

Hi all!
@st3iny thanks! Let me know what do you think about implementation! 

I haven't tested it for a long time, and I hope that everything works as before. In any case, let me know how you test it.

st3iny commented:
> > @sazanof Tests have to be updated for the new `attachments` property.
> 
> Tell me, is it enough to add a parameter or do I need to write a full test? If it's a complete test, then I ask for help, since **I've never written tests and don't understand how to write them** (It's a matter of time and effort, I don't argue). =( I tried to write my own block for checking (by analogy with other tests) `it()` This error occurs if I add my parameters to
> 
> ```
> expect(getDefaultEvent Object()).to Equal({
> ....
> attachments: [
> 	'ATTACHMENT1',
> 	'ATTACHMENT2',
> 	'ATTACHMENT3',
> 	'ATTACHMENT4',
> 	]
> })
> ```
> 
> https://gist.github.com/sazanof/ecb2aaa8c760f93e74d2952294ba09ba
> 
> When checking, I get an error. I can't understand why it occurs...
> 
> ```
>  ● Test suite failed to run
> 
>     TypeError: Converting circular structure to JSON
>         --> --> --> --> --> --> --> --> starting at object with constructor 'CalendarComponent'
>         --- property '_root' closes the circle
>         at stringify (<anonymous>)
> ```

Maybe a bit late: Anything you put into `expect(...)` has to a plain object that can be converted to JSON. Most likely, you are trying to pass an object that has some form of reference to itself (e. g. via child -> parent -> child -> parent). It probably is caused by `expect(mapEventComponentToEventObject(eventComponent))`.

st3iny reviewed (commented):
Tested and works so far.

### I noticed multiple small issues:
1. ~Share links are not deleted once an attachment is removed. The file should be unshared in this case.~
2. ~Share links are not deleted once an event has no attendees anymore.~
3. Attachments are not saved into a sub folder (e. g. `/Calendar`). Instead, they are saved to the root directory which can get polluted quite quickly.
4. Uploaded attachments are silently overwritten if they have the same name. Example: Upload an attachment, edit another event and upload a similar file with the same name again -> The first attachment file will be overwritten without any warning.

I'm not sure what to do about 4. Maybe just append a random hash of 

**EDIT:** Regarding 1 and 2: This makes sense for recurring events because there might be recurrences that still have attendees/attachments. However, for regular events this might be worth to consider. I'm fine with ignoring this for now.

**EDIT 2:** Regarding 4: I just tested this in Talk (as they are essentially doing the same thing) and they just append a suffix (e. g. `some file (2).png`) to the file name.

sazanof commented:
@st3iny, thnaks! I have a question:

> Attachments are not saved into a sub folder (e. g. /Calendar). Instead, they are saved to the root directory which can get polluted quite quickly.

Did we decide that it would be a subfolder `/Calendar`?
And if so, how to create it? Probably this should happen at the time of activation or installation of the application itself?

And a completely noob question... how do I pull up the changes that you have made?
`git pull origin enh/attachements-to-events` -> Already updated =(

st3iny commented:
> @st3iny, thnaks! I have a question:
> 
> > Attachments are not saved into a sub folder (e. g. /Calendar). Instead, they are saved to the root directory which can get polluted quite quickly.
> 
> Did we decide that it would be a subfolder `/Calendar`? And if so, how to create it? Probably this should happen at the time of activation or installation of the application itself?

You can create folders using axios and the DAV API. You could create the folder just before uploading the attachment. If it already exist, the API will return status code 405.

```javascript
axios({
	method: 'MKCOL',
	url: `/remote.php/dav/files/${dav.userId}/Calendar`,
}).
```

> And a completely noob question... how do I pull up the changes that you have made? `git pull origin enh/attachements-to-events` -> Already updated =(

Hmmm ... git pull should just work. You could try `git pull --rebase origin enh/attachements-to-events`. If this doesn't work either and you have no uncommitted and/or unpushed local changes you could try `git fetch && git reset --hard origin/enh/attachements`. **Warning:** This will delete all local changes on this branch (if any).

sazanof commented:
> You can create folders using axios and the DAV API. You could create the folder just before uploading the attachment. If it already exist, the API will return status code 405.

OK, let`s check

BTW, There is an idea to ask the user whether he wants to create a duplicate file or overwrite it or not.

szaimen commented on src/services/attachmentService.js:
```diff
@@ -3,6 +3,8 @@ import { generateOcsUrl } from '@nextcloud/router'
 import { showError, showSuccess } from '@nextcloud/dialogs'
 import { translate as t } from '@nextcloud/l10n'
 
+const FOLDER = 'Calendar'
```
can we somehow make the folder configurable? I know a few cases where this folder will not work in the user root. (e.g. when it is read only). 

E.g. this might be a good example: https://github.com/nextcloud/photos/blob/master/src/components/Settings/PhotosLocationSettings.vue

szaimen replied:
Ping @sazanof ? :)

sazanof replied:
> Ping @sazanof ? :)
```
PING @sazanof  (127.0.0.1) 56(84) bytes of data.
64 bytes from @sazanof  (127.0.0.1): icmp_seq=1 ttl=247 time=500000 ms
```
Yes, I looked at how it is implemented in the Photo application. In fact, we can add routes on the backend and expand the configuration on the frontend. I'll do it! :100: 

PS: Pings are unpredictable now.

sazanof replied:
https://github.com/nextcloud/calendar/pull/4251/commits/ee408dc0a974625b8320203e1db4e16d21ccb1b8

Well, it was difficult, but that's in the past.

szaimen commented on src/components/Editor/Attachments/AttachmentsList.vue:
```diff
@@ -0,0 +1,241 @@
+</template>
+
+<script>
+import ListItem from '@nextcloud/vue/dist/Components/NcListItem'
```
i think the imports are probably missing the file endings...

sazanof replied:
Done via [e532920](https://github.com/nextcloud/calendar/pull/4251/commits/e532920b514c965edbd96e6138dafbaeba82eb52)

`150 problems (0 errors, 150 warnings)` =)

tcitworld commented on lib/Controller/SettingsController.php:
```diff
@@ -172,6 +174,28 @@ private function setShowTasks(string $value):JSONResponse {
+		try {
+			$this->config->setUserValue(
+				$this->userId,
+				$this->appName,
```
Once the attachment part will be handled on the server by [Managed Attachments](https://datatracker.ietf.org/doc/html/rfc8607), it would be nice to keep this setting value, so we could set the preference in `dav` instead of `calendar` (`$this->appName`).

sazanof replied:
Simply change to 

```
$this->config->setUserValue(
				$this->userId,
				'dav',
				'attachmentsFolder',
				$value
			);
			```

?

sazanof replied:
https://github.com/nextcloud/calendar/pull/4251/commits/5a5638c6ff7a1b7d336b1fc8a90b913a8d32bb8b

tcitworld commented on lib/Controller/ViewController.php:
```diff
@@ -100,6 +100,7 @@ public function index():TemplateResponse {
 		$showWeekNumbers = $this->config->getUserValue($this->userId, $this->appName, 'showWeekNr', $defaultWeekNumbers) === 'yes';
 		$skipPopover = $this->config->getUserValue($this->userId, $this->appName, 'skipPopover', $defaultSkipPopover) === 'yes';
 		$timezone = $this->config->getUserValue($this->userId, $this->appName, 'timezone', $defaultTimezone);
+		$attachmentsFolder = $this->config->getUserValue($this->userId, $this->appName, 'attachmentsFolder', '/');
```
Save as above, use `dav` instead of `calendar`

sazanof replied:
https://github.com/nextcloud/calendar/pull/4251/commits/5a5638c6ff7a1b7d336b1fc8a90b913a8d32bb8b

tcitworld commented on lib/Controller/ViewController.php:
```diff
@@ -100,6 +100,7 @@ public function index():TemplateResponse {
 		$showWeekNumbers = $this->config->getUserValue($this->userId, $this->appName, 'showWeekNr', $defaultWeekNumbers) === 'yes';
 		$skipPopover = $this->config->getUserValue($this->userId, $this->appName, 'skipPopover', $defaultSkipPopover) === 'yes';
 		$timezone = $this->config->getUserValue($this->userId, $this->appName, 'timezone', $defaultTimezone);
+		$attachmentsFolder = $this->config->getUserValue($this->userId, $this->appName, 'attachmentsFolder', '/');
```
Default to `/calendar`, not the root folder?

sazanof replied:
I did so because the /Calendar folder is missing by default. And deleted the code that creates the /calendar folder before uploading local attachments. any ideas how to create folder once?

st3iny replied:
Checking for the folder each time involves a network request, too. I suggest checking the folder once on opening the calendar app and then store the boolean. The MKCOL operation can be done once an attachment file needs to be created if the folder does not exist yet.

sazanof replied:
Hello! So, you can test this. I updated the code.

tcitworld commented:
Just noticed in https://datatracker.ietf.org/doc/html/rfc8607#section-4.3
> This property (`MANAGED-ID`) parameter MUST NOT be present in the case of unmanaged attachments.

So I'm no longer sure we should use it before implementing Managed Attachments in server. :thinking:

tcitworld commented:
Another few notes from reading [RFC 8607](https://datatracker.ietf.org/doc/html/rfc8607):
* Changing attachments should trigger scheduling messages to attendees https://datatracker.ietf.org/doc/html/rfc8607#section-3.12.6
* `MANAGED-ID` occurrences must be removed when exporting and importing data https://datatracker.ietf.org/doc/html/rfc8607#section-3.12.7

sazanof commented:
`Ping @all` =)

I'm thinking is there an option to create (check) a user folder without adding an additional queries of the form? (I hope I wrote it right. Translation difficulties, you know=))

```
axios({
	method: 'MKCOL',
	url: `/remote.php/dav/files/${dav.userId}/Calendar`,
}).

```
Who thinks what? 

see https://github.com/nextcloud/calendar/pull/4251#discussion_r1011385909

_There is very little left =)_

miaulalala commented:
> Another few notes from reading [RFC 8607](https://datatracker.ietf.org/doc/html/rfc8607):
> 
> * Changing attachments should trigger scheduling messages to attendees https://datatracker.ietf.org/doc/html/rfc8607#section-3.12.6

This should be doable by hooking into the sabre event lifecycle methinks.

> * `MANAGED-ID` occurrences must be removed when exporting and importing data https://datatracker.ietf.org/doc/html/rfc8607#section-3.12.7

I think the export should be ok, we sanitise the data a lot. Of course testing is needed. 
The import might work or it might not - `\Sabre\VObject\Component\VEvent::getValidationRules` is pretty strict, had this in another ticket where the (valid) ENCODING param threw.

st3iny reviewed (commented):
I tested it again and it works really great now! I fixed the share creation as it wasn't using the attachment folder.

There is a bug with the attachment folder picker and I'm currently researching what is causing it.

sazanof commented:
Hello, @st3iny! Thank you for review =)
Hello everyone! So... drum roll...
> Tested and works!
> 
> Please squash **all** commits (including mine) into a single commit and then we are ready to merge.

szaimen commented:
🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉

sazanof commented:
*** Sounds of a champagne cork flying out ***

ChristophWurst commented:
@sazanof https://youtu.be/CeKNPSsK-kI?t=524 :pray:
